lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+            hedges.sort!
+            if hedges.size > 1
+              (0..hedges.size - 2).each do |i|
+                e1 = hedges[i]
+                e2 = hedges[i + 1]
+
+                e1.prev = e2.twin
+                e2.twin.next = e1
+              end
+
+              hedges[-1].prev = hedges[0].twin
+              hedges[0].twin.next = hedges[-1]
+
+            # handle case of dangling line
+            else
+              e = hedges[0]
+              e.twin.next = e
+              e.prev = e.twin
+            end
+          end
+        end
+
+        # It is possible that intersections occur when new edges are added.
+        # This will split those edges into more half-edges while preserving
+        # the existing half-edges when possible since geometries may reference
+        # them.
+        def compute_split_edges(seg, ints)
+          points = ints.uniq.sort do |a, b|
+            a.distance(seg.s) <=> b.distance(seg.s)
+          end
+          points = [seg.s] + points + [seg.e]
+
+          he_start = incident_edges[points.first.coordinates].find { |he| he.destination == points.last }
+          he_end = incident_edges[points.last.coordinates].find { |he| he.destination == points.first }
+
+          points.each_cons(2) do |s, e|
+            edge = Segment.new(s, e)
+            if s == he_start.origin
+              he = HalfEdge.new(e)
+              he_start.twin = he
+              he.twin = he_start
+              insert_half_edge(he)
+            elsif e == he_end.origin
+              he = HalfEdge.new(s)
+              he_end.twin = he
+              he.twin = he_end
+              insert_half_edge(he)
+            else
+              create_half_edge(edge)
+            end
+            @edges << edge
+          end
+          @edges.delete(seg)
+        end
+      end
+
+      # GeometryGraph is a PlanarGraph that is built by adding
+      # geometries instead of edges. The GeometryGraph will
+      # hold a reference to an arbitrary HalfEdge on the
+      # interior of the geometry for every boundary in the geometry.
+      # For example, a polygon will have a reference to a HalfEdge for its
+      # exterior shell and one for every hole.
+      class GeometryGraph < PlanarGraph
+        # GeomEdge will be used to store the references to the HalfEdges
+        GeomEdge = Struct.new(:exterior_edge, :interior_edges)
+
+        def initialize(geom)
+          super()
+          @parent_geometry = geom
+          @geom_edges = []
+          add_geometry(geom)
+        end
+        attr_reader :parent_geometry, :geom_edges
+
+        private
+
+        # Adds a geometry to the graph and finds its
+        # reference HalfEdge(s).
+        #
+        # @param geom [RGeo::Feature::Instance]
+        def add_geometry(geom)
+          case geom
+          when Feature::Point
+            # Can't handle points yet, so just add an empty entry for them
+            @geom_edges << GeomEdge.new
```

Comment:
I guess we'll see in the next PR that we don't use those for points, right?

Reply:
Yes, but we need to at least have a blank entry in the `@geom_edges` array so that we can properly access each geometry in a collection.